src/branch_and_bound.rs

```diff
@@ -1,222 +1,785 @@
+        //
+        // That is, the range includes solutions that exactly equal the target up to but not
+        // including values greater than target + cost_of_change.
+        //
+        // if current_waste > best_waste, then backtrack.  However, only backtrack if
+        // it's high fee_rate environment.  During low fee environments, a utxo may
+        // have negative waste, therefore adding more utxos in such an environment
+        // may still result in reduced waste.
+        else if value > upper_bound || current_waste > best_waste && fee_rate > long_term_fee_rate
+        {
+            backtrack = true;
+        }
+        // * value meets or exceeds the target.
+        //   Record the solution and the waste then continue.
+        else if value >= target {
+            backtrack = true;
+
+            let v = value.to_signed().ok()?;
+            let t = target.to_signed().ok()?;
+            let waste: SignedAmount = v.checked_sub(t)?;
+            current_waste = current_waste.checked_add(waste)?;
+
+            // Check if index_selection is better than the previous known best, and
+            // update best_selection accordingly.
+            if current_waste <= best_waste {
+                best_selection = Some(index_selection.clone());
+                best_waste = current_waste;
+            }
 
-fn find_solution<T: Utxo>(
-    target: u64,
-    cost_of_change: u64,
-    utxo_pool: &mut [T],
-) -> Option<Vec<bool>> {
-    let utxo_sum = utxo_pool.iter().fold(0u64, |mut s, u| {
-        s += u.get_value();
-        s
-    });
+            current_waste = current_waste.checked_sub(waste)?;
+        }
 
-    let utxo_pool_length = utxo_pool.len();
-    utxo_pool.sort_by_key(|u| Reverse(u.get_value()));
+        // * Backtrack one node
+        if backtrack {
+            let last_index = index_selection.pop().unwrap();
+            let (eff_value, utxo_waste, _) = w_utxos[last_index];
 
-    let mut curr_selection: Vec<bool> = vec![false; utxo_pool_length];
-    let mut best_selection = None;
-    let mut remainder = utxo_sum;
+            current_waste = current_waste.checked_sub(utxo_waste)?;
+            value = value.checked_sub(eff_value)?;
+            index -= 1;
+            assert_eq!(index, last_index);
+        }
+        // * Backtrack to new tree
+        else if backtrack_subtree {
+            // No new subtree left to explore.
+            if index_selection.is_empty() {
+                return index_to_utxo_list(best_selection, w_utxos);
+            }
 
-    let lower_bound = target;
-    let upper_bound = cost_of_change + lower_bound;
+            // Anchor the new subtree at the next available index.
+            //
+            // if our index selection is: [0,1,2,3]
+            // then copy the head 0 into index.
+            // At the end of this loop, index is incremented to 1.
+            // Therefore, we will be starting our next search tree at index 1.
```

Comment:
It sounds to me like you are using a different order of exploration than Bitcoin Core’s BnB implementation. Bitcoin Core uses a depth-first search, the traversed UTXO combinations of four UTXOs 1, 2, 3, and 4 would be (underscore is an omitted UTXO):

```
{} {1} {12} {123} {1234} {123_} {12_} {12_4} {12__} {1_} {1_3} {1_34} {1_3_} {1__} {1__4} {1___} {_} {_2} {_23} {_234} {_23_} {_2_} {_2_4} {_2__} {__} {__3} {__34} {__3} {___} {___4} {____}
```

I’ve described the same search pattern in a lot of detail here on the CoinGrinder PR: https://github.com/bitcoin/bitcoin/pull/27877/commits/6cc9a46cd0f4ed80d4523bbef1508142e0c80d27

I’m not particularly familiar with Rust, and it’s not obvious to me whether the search strategy employed here will enumerate all possible combinations. In case you want to verify that, I would suggest that you print each evaluated input combination your search traverses to check whether all input combinations that should be enumerated get traversed. I suspect that immediately jumping to the next index as described in the comment here would skip part of your search tree.

Reply:
As with bitcoin core bnb, the first thing that happens is a reverse sort: https://github.com/bitcoin/bitcoin/blob/master/src/wallet/coinselection.cpp#L114.   four UTXO combinations 1,2,3,4 would be searched depth first starting at 4.

`{} {4} {43} {432} {4321}`

If a print statement is add here: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/blob/feature/iter-branch-and-bound/src/branch_and_bound.rs#L318 where the next depth is explored (depth + 1).

`println!("{}", eff_value)`

The results will be 4,3,2,1 if the test for 10 is run with the following:

`cargo test select_coins_bnb_ten -- --nocapture`

Note that `-- --nocapture` is needed when running the test suit to print to stdout.

I'm struggling to find other samples that would prove this is a depth first search.  In this case, the search stops after 4,3,2,1 because once 4 is remove from the inclusion branch, there is no longer enough value left to reach 10, so the search ends.

> I’m not particularly familiar with Rust

If there's anything about rust that is making your review slow or not possible I'm happy to help answer any questions.

Reply:
One more less trival example five to show this matches what bitcoin-core would find.

Placing a print statement at L318, will print:
`4,3,2,1,3,2`

that is:
`{} {4} {4,3} {4,2} {4,1} {3} {3,2}`

This is the same DFS and set evaluation that core performs and does not skip any part of the search tree.

Reply:
run `cargo test select_coins_bnb_five -- --nocapture` to duplicate

Reply:
Given UTXOs with the effective values { 10, 7.00000005, 6.00000005, 6, 3, 2, 1.00000005 } and a target of 18, what is the best solution and what does your implementation return?

Reply:
I added a test (in the quickest possible way for now) to this branch https://github.com/yancyribbens/rust-bitcoin-coin-selection/tree/test/add-extended-tests.

see commit: https://github.com/yancyribbens/rust-bitcoin-coin-selection/commit/b46e8bd1e7ee4f88ad80d0834bedd580870de416

Using 50 sats for a cost_of_change:
the implementation returns: `6.00000005, 6, 3, 2, 1.00000005`

Happy to flush out more tests in detail, but just wanted to reply with this to show that the implementation does work..

Reply:
oops sorry, see this commit https://github.com/yancyribbens/rust-bitcoin-coin-selection/commit/fcd1908295821423839254e7e596c55bb263f53c

Reply:
On further inspection, and drawing out a number of binary tree diagrams, I do think this misses some solutions because of the way I implemented backtracking with backtrack subtree.  That's embarrassing but hopefully fixing the backtrack won't be too painful.  I'll take a look at Coin Grinder also for some inspiration.

Reply:
>  what is the best solution and what does your implementation return?

I think the best solution is: [7.00000005, 6, 3, 2] = 18.00000005

My implementation returns [6.00000005, 6, 3, 2, 1.00000005] = 18.0000001

Thanks for providing a better this test case.  I see now that once at a leaf node, there is another backtrack case where we don't backtrack once and we don't backtrack all the way back and omit the root, but another case in between the two.

Reply:
What I meant is, there is another backtrack case where we don't backtrack once and we don't backtrack all the way back and omit the root, but another case in between the two.

Reply:
The best solution is `{10, 6, 2}`.

I feel like I should explain why I told you that I don’t have time for your project. I am frustrated because I feel that I have provided numerous resources to help understand the algorithm, and am kept repeating myself at this point. You should address this project’s issues at the root by implementing adequate test coverage and taking the time to _fully understand_ what you are trying to implement.

My experience is that when I try to point you in a direction that should lead to you _discovering_ and _understanding_ what the problem is, it feels to me like you are putting in too little effort to understanding my feedback and instead of pondering why I write what I suggest and following my advice, arrive at uncharitable interpretations like I don’t understand something as obvious as your UTXOs being sorted by descending effective value and immediately ask for more help. So, on the one hand, you are asking me to invest my time, guilt-trip me by saying that you don’t have any other reviewer, but on the other hand, you make me feel that you seem to think that you already know better than me anyway.

> Thanks for providing a better this test case. I see now that once at a leaf node, there is another backtrack case where we don't backtrack once and we don't backtrack all the way back and omit the root, but another case in between the two.

As I [wrote last week](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/30#discussion_r1500685154): 

> Following this advice would help you understand why I don’t use something like backtrack_subtree, but instead [backtrack twice](https://github.com/bitcoin/bitcoin/blob/1ac627c485a43e50a9a49baddce186ee3ad4daad/src/wallet/coinselection.cpp#L402-L411). I would expect that it would be helpful to investigate what issues are covered by the BnB tests in Bitcoin Core. You may want to replicate more of those tests. To that end, it may be useful to look at the old tests and my reimplementation: [diff](https://github.com/bitcoin/bitcoin/compare/e6dbf48c64a3c267425911d768a77b95a6c3d078..b27e06608fe9f58a2a6377547738758a7d0bbe7c), [list of commits in the reimplementation](https://github.com/bitcoin/bitcoin/pull/28985/commits)

You have to backtrack twice. I.e. if your selection out of the UTXOs {ABCDEFG} was {ABDEF} you need to deselect the last {ABDEF} ⇒ {ABDE}, then move the remaining last selected one down: {ABDE} ⇒ {ABDF}. This is explained at least twice in the CoinGrinder documentation, which is why I linked you to it, and should become apparent if you think about the order of all leaves in a depth-first search and play through which ones you can skip but which ones you have to visit e.g. on a whiteboard. 

I really appreciate your interest in coin selection and how long you have been sticking with this project, but I  would appreciate if I saw more proactive effort in comprehending my feedback instead of making me repeat myself, because it is making me feel that you do not respect my time.

Reply:
By the way, in case that was not clear: CoinGrinder is a reimplementation of _exactly_ the same algorithm as BnB except that there is neither excess, nor an upper bound for the target, because CoinGrinder creates a change output.

Reply:
My intent was not to guilt trip you into reviewing.  Apologies if that's how I came off.  It's just that in my opinion, you wrote the book on coin-selection and are the expert which is why I seek your input and review.  I feel like this implementation is much better than my first attempt, but as you pointed out still anemic in how the backtrack is currently working.  I tried to get others to look at it and review it, but none could really follow the logic in bitcoin core well enough to give me feedback on the most critical component.  To be fair I made a blunder by trying to optimize what I thought was the behavior instead of more dutifully trying to copy the semantics which I take responsibility for.  Anyway, thanks as for providing feedback and if there is a future for working together, I will try to more diligently ponder your advice, feedback and provided resources before responding with questions/assumptions.  Also, thanks for explaining more about your frustration and where you are coming from.  You saved me making another mistake with this implementation by reviewing this before it was merged.  I'm not sure how to repay you but consider me in your debt.

Reply:
Thank you for your understanding. Yes, I do note that this implementation is much better than the first.

In my experience, trying to optimize early or taking shortcuts tends to break things in ways that are hard to diagnose more often than not. While it may feel slower while you are doing it, I find that I’m overall much faster when I take the time to first make the simplest possible implementation, ensure that it is correct, and only then consider whether it needs to be made more efficient. If it needs to be improved, I do one improvement at a time, so I know what I need to test and what change broke the correctness if any test fails.

Some things that helped me over time:

- Commit early, commit often
- Slow is smooth, smooth is fast
- Assume that every untested aspect of your implementation is broken

You can probably make your tests much easier to extend and maintain by creating some helper methods, check out for example `TestBnBSuccess` here: https://github.com/bitcoin/bitcoin/pull/28985/commits/4a95b234918e9f714ac6f707defe17a777225e47#diff-36088c93368e137d955348aba223985bd4f198f2aaecd626c830f4612ca884c8R100-R133